docs: add cryptography documentation (#280)
canonical/craft-store: docs/reference/cryptography.rst

Author: Ah, I didn't move the file! Sorry about that.

.. _reference_cryptography:

Cryptographic Documentation
===========================

Craft Store uses cryptographic processes to send files to and from the internet. It does
not directly implement its own cryptography, but it does depend on external libraries to
do so.

Authentication
--------------

> Craft Store handles repeated authentication via macaroons, which are processed by the

Reviewer: Given that macaroons are not exactly widely known, maybe a link would be valuable?

That said, IDK what link to put there. [This?](https://research.google/pubs/macaroons-cookies-with-contextual-caveats-for-decentralized-authorization-in-the-cloud/) Just a [Wikipedia link](https://en.wikipedia.org/wiki/Macaroons_(computer_science))?

Author: I wanted to do the same but I agree, they're not well documented. I would say the two links you sent are just about the only two I've been able to find talking about them in a (reasonably) user-friendly way. @medubelko which of the two, if either, do you think fits best in our docs?

Reviewer 2: Let's go with the Google page.

`macaroonbakery <https://pypi.org/project/macaroonbakery/>`_ library. This library
validates and manages macaroons as returned by storefronts and simplifies the inclusion
of macaroons in further requests to storefronts.

Credentials may additionally be stored on-disk using the `keyring
<https://pypi.org/project/keyring/>`_ library, which will use the keyring service
provided by the host operating system. If the host OS does not have a keyring service,

> they will instead be stored in plain text. A warning is issued to the terminal when this

Reviewer: ```suggestion
they will instead be stored in a plain text called `credentials.json` under the application's data data storage directory. A warning is issued to the terminal when this
```

.. _reference_cryptography:

Cryptographic Documentation
===========================

> Craft Store uses cryptographic processes to send files to and from the internet. It does

Reviewer 2: > to and from the internet

I don't feel confident about this phrase. What about "between devices and endpoints through the internet"?

Author: I'm going to slim it down to just "endpoints" because it should only really ever be communicating directly with one of the stores, but otherwise I agree with this.

Author: Nevermind, I realize now that by "devices" I think you mean the machine executing the library. I'll keep "devices" in there too

Reviewer 2: "Devices" is a common shorthand for "end user computing devices", so if it _doesn't_ involve the user's machine, then we should leave it out. :)

not directly implement its own cryptography, but it does depend on external libraries to
do so.

Authentication
--------------

> Craft Store handles repeated authentication via macaroons, which are processed by the

Reviewer 2: > handle repeated authentication

I get what this means, but it sounds awkward and potentially confusing to me. Is this a fixed phrase, or could we consider something like "is an authentication provider"?

Author: Not fixed at all - I like your phrasing much more. Thanks!

`macaroonbakery <https://pypi.org/project/macaroonbakery/>`_ library. This library

> validates and manages macaroons as returned by storefronts and simplifies the inclusion

Reviewer 2: > storefronts

I know this is a term, but do you think "store front-ends" would be clearer?

Author: Hmm. I'm not sure - "storefronts" comes naturally to me, but I don't know if we have a precedent anywhere. I feel like we'd usually refer to it as simply "the store" when discussing, say, the Snap Store, but that definitely doesn't fit.

If you have a strong opinion, I can change to "store front-ends". Some other options, for your consideration:
- store services
- store providers
- storefront servers
- stores

Reviewer 2: I'd be fine with "stores" – I use that in Snapcraft to cover both public and private stores – but I figured we were trying to be more precise. I made the note about endpoints because I thought we were distinguishing FE from BE.

> of macaroons in further requests to storefronts.

Reviewer 2: ```suggestion
of macaroons in further requests to stores.
```

Credentials may additionally be stored on-disk using the `keyring
<https://pypi.org/project/keyring/>`_ library, which will use the keyring service

> provided by the host operating system. If the host OS does not have a keyring service,

Reviewer 2: ```suggestion
provided by the host operating system. If the host does not have a keyring service,
```

they will instead be stored in a plain text file called :file:`credentials.json` under
the application's data storage directory. A warning is issued to the terminal when this
behavior is triggered. This behavior is available to ease the usage of Craft Store
within CI containers, but is generally discouraged.

> Network connections
> -------------------

Reviewer 2: A nitpick:

```suggestion
Network connectivity
--------------------
```

Craft Store handles URLs using `urllib
<https://docs.python.org/3/library/urllib.html>`_. The use of this library both
simplifies and hardens the parsing of URLs provided by consuming applications.

> Connections over the internet are done by the `requests

Reviewer 2: ```suggestion
Connections over the internet are mediated by the `requests
```

<https://requests.readthedocs.io/en/latest/>`_ or `httpx
<https://www.python-httpx.org/>`_ libraries, which are configured to always
attempt HTTPS connections first, but have the ability to communicate over HTTP
as a fallback. Canonical storefronts do not support HTTP, but this capability

> is kept in to aid with local testing. The distinction of which library is used

Reviewer 2: ```suggestion
is retained to aid with local testing. The distinction of which library is used
```

> depends on the request object provided by an application using this library.

Reviewer 2: > request object

I'm assuming we mean an object in the sense from the HTTP protocol? HTTP likes to borrow technical terms and use them in new ways. Can we link to the MDN page about this to help clarify?

Author: Actually, it's a secret other thing: a Python object either from the `requests` or `httpx` library. How about this:

> The distinction of which library is used depends on the HTTP request backend passed to Craft Store ...

> depends on the request object provided by an application using this library.

Reviewer 2: > The distinction of which library... depends on... this library.

This reads like circular logic to me. Is there a way we can simplify this sentence?

Author: Combining this with the previous comment, I think this might be good, but I'd like your feedback:

> Between these two libraries, Craft Store will use whichever of the two it is invoked with by the consuming application.

Reviewer 2: That works! And let's remove the redundant words:

> Between these two libraries, Craft Store will use whichever of the two is invoked by the consuming application.

These libraries handle cryptographic operations such as the TLS handshake that

> are standard requirements for modern internet communication.

Reviewer 2: This last sentence feels like it belongs as the second sentence in the paragraph.
